libr/search/search.c

```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
-				char ch = kw->bin_keyword[kw->idx[j]]; // signed char
-				ut8 ch2 = buf[i];
-				/* no icase in delta keys */
-				/* no binmask in delta keys */
-				/* no inverse support for delta keys */
-				if (pch+ch == ch2) {
-					kw->idx[j]++;
-					if (kw->idx[j] == kw->keyword_length) {
-						if (!r_search_hit_new (s, kw, (ut64)
-							from + i - kw->keyword_length + 1))
-							return -1;
-						kw->idx[j] = 0;
-						//kw->idx[0] = 0;
-						kw->distance = 0;
-						kw->count++;
-						count++;
-						//s->nhits++;
+	r_list_foreach (s->kws, iter, kw) {
+		ut8 *a = kw->bin_keyword;
+		for (i = 0; i + kw->keyword_length < len; i++) {
+			if ((ut8)(buf[i+1] - buf[i]) == a[0]) {
+				for (j = 1; j < kw->keyword_length &&
+							 (ut8)(buf[i+j+1] - buf[i+j]) == a[j]; j++) {}
+				if (j == kw->keyword_length) {
+					if (!r_search_hit_new (s, kw, from + i)) {
+						return -1;
+					}
+					kw->count++;
+					count++;
+					if (++s->nhits >= s->maxhits) {
+						return count;
+					}
+					if (!s->overlap) {
+						i += kw->keyword_length;
 					}
 				}
-				pch = ch2;
 			}
 		}
 		count = 0;
 	}
 	return count;
 }
 
-/* Boyer-Moore-Horspool pattern matching */
-#if USE_BMH
-R_API int r_search_bmh (const RSearchKeyword *kw, const ut64 from, const ut8 *buf, const int len, ut64 *out) {
+// Boyer-Moore-Horspool pattern matching
+// Supported search variants: icase, overlap
+static int r_search_horspool(RSearch *s, RSearchKeyword *kw, ut64 from, const ut8 *buf, int len) {
 	ut64 bad_char_shift[UT8_MAX + 1];
-	ut64 pos = from;
-	int i, kw_len;
+	int i, j, m = kw->keyword_length - 1, count = 0;
 	ut8 ch;
 
-	kw_len = kw->keyword_length - 1;
-
-	if (kw_len < 0)
-		return false;
-
-	for (i = 0; i < 256; i++)
+	for (i = 0; i < R_ARRAY_SIZE (bad_char_shift); i++)
 		bad_char_shift[i] = kw->keyword_length;
-
-	for (i = 0; i < kw_len; i++) {
+	for (i = 0; i < m; i++) {
 		ch = kw->bin_keyword[i];
-		bad_char_shift[kw->icase?tolower(ch):ch] = kw_len - i;
+		bad_char_shift[kw->icase ? tolower (ch) : ch] = m - i;
 	}
 
-	while (pos < len) {
-		for (i = kw_len; ; i--) {
-			ut8 ch1 = buf[pos + i];
-			ut8 ch2 = kw->bin_keyword[i];
+	for (i = 0; i < len; ) {
+		for (j = m; ; j--) {
+			ut8 a = buf[i + j], b = kw->bin_keyword[j];
 			if (kw->icase) {
-				ch1 = tolower(ch1);
-				ch2 = tolower(ch2);
+				a = tolower (a);
+				b = tolower (b);
 			}
-			if (kw->binmask_length && i < kw->binmask_length) {
-				ch1 &= kw->bin_binmask[i];
-				ch2 &= kw->bin_binmask[i];
-			}
-			if (ch1 != ch2)
+			if (a != b)
 				break;
 			if (i == 0) {
-				if (out)
-					*out = pos;
-				return true;
+				if (!r_search_hit_new (s, kw, from + i)) {
+					return -1;
+				}
+				kw->count++;
+				count++;
+				if (!s->overlap) {
+					i += kw->keyword_length;
+					goto next;
+				}
 			}
 		}
-		ch = buf[pos + kw_len];
-		pos += bad_char_shift[kw->icase?tolower(ch):ch];
+		ch = buf[i + m];
+		i += bad_char_shift[kw->icase ? tolower (ch) : ch];
+	next:;
```

Comment:
it will be probably easier to read if you put this label at the begining of the loop. also we may assume i>0 otherwise this can result in an infinite loop.